ori/volume/server/server.go

```diff
@@ -17,66 +17,72 @@ package server
 import (
 	"context"
 	"fmt"
+	"os"
+	"sync"
 
 	"github.com/go-logr/logr"
-	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
-	volumev1alpha1 "github.com/onmetal/cephlet/ori/volume/api/v1alpha1"
-	"github.com/onmetal/cephlet/ori/volume/apiutils"
-	computev1alpha1 "github.com/onmetal/onmetal-api/api/compute/v1alpha1"
-	ipamv1alpha1 "github.com/onmetal/onmetal-api/api/ipam/v1alpha1"
-	networkingv1alpha1 "github.com/onmetal/onmetal-api/api/networking/v1alpha1"
-	storagev1alpha1 "github.com/onmetal/onmetal-api/api/storage/v1alpha1"
 	"github.com/onmetal/onmetal-api/broker/common/cleaner"
 	"github.com/onmetal/onmetal-api/broker/common/idgen"
 	ori "github.com/onmetal/onmetal-api/ori/apis/volume/v1alpha1"
-	corev1 "k8s.io/api/core/v1"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/runtime/schema"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
-	kubernetes "k8s.io/client-go/kubernetes/scheme"
-	"k8s.io/client-go/rest"
+	"golang.org/x/sync/singleflight"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/yaml"
 	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 )
 
-var scheme = runtime.NewScheme()
+type Provisioner interface {
+	CreateOsImage(ctx context.Context, imageName, imageId string) error
+	GetOsImage(ctx context.Context, imageName string) (string, bool, error)
+	DeleteOsImage(ctx context.Context, imageName, imageId string) error
 
-func init() {
-	utilruntime.Must(kubernetes.AddToScheme(scheme))
-	utilruntime.Must(computev1alpha1.AddToScheme(scheme))
-	utilruntime.Must(networkingv1alpha1.AddToScheme(scheme))
-	utilruntime.Must(storagev1alpha1.AddToScheme(scheme))
-	utilruntime.Must(ipamv1alpha1.AddToScheme(scheme))
-	utilruntime.Must(snapshotv1.AddToScheme(scheme))
+	CreateCephImage(ctx context.Context, imageId string, volume *ori.Volume, class *ori.VolumeClass, osImageId string) (*CephImage, error)
+	GetCephImage(ctx context.Context, imageId string) (*CephImage, error)
+	ListCephImages(ctx context.Context) ([]*CephImage, error)
+	DeleteCephImage(ctx context.Context, imageId string) error
+
+	FetchAuth(ctx context.Context) (string, string, error)
+
+	Monitors() string
 }
 
 type Server struct {
-	client client.Client
-	idGen  idgen.IDGen
+	idGen       idgen.IDGen
+	provisioner Provisioner
+
+	inProgress     sets.Set[string]
+	inProgressLock sync.Mutex
+
+	VolumeNameLabelName    string
+	AvailableVolumeClasses map[string]ori.VolumeClass
 
-	namespace string
+	syncPopulation singleflight.Group
+}
+
+func (s *Server) lock(volumeName string) error {
```

Comment:
Why does this return an `error`? It has to prevent multiple operations from operating on the same volume, yes, but it shouldn't error and rather let the subsequent callers wait until the previous lock has been released. You could (probably, don't pin me on that) achieve that using `sync.Cond` / to a certain degree even with the `singleflight.Group` (depending on impementation ofc).

Reply:
I think it's fine to return an error because of the following reasons: 
- we are locking based on a ceph image id. In case of volume creation it will only return an error if we have selected a duplicate  UUID
- in case of operating on the same image at the same time (like deleting a `Volume` which is not fully provisioned): that's not allowed and we can do an early return. The caller should do a retry